File: SenLib/Sen4/FileFixes/t_mstqrt.cs
using HyoutaUtils;
using HyoutaUtils.Streams;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace SenLib.Sen4.FileFixes {
	class MasterQuartzMemo {
		public ushort mqidx;
		public ushort stridx;
		public string str;

		internal MasterQuartzMemo(byte[] data) {
			var stream = new DuplicatableByteArrayStream(data);
			mqidx = stream.ReadUInt16();
			stridx = stream.ReadUInt16();
			str = stream.ReadUTF8Nullterm();
		}

		internal byte[] ToBinary() {
			MemoryStream ms = new MemoryStream();
			ms.WriteUInt16(mqidx);
			ms.WriteUInt16(stridx);
			ms.WriteUTF8Nullterm(str);
			return ms.CopyToByteArrayAndDispose();
		}
	}

	class t_mstqrt : FileMod {
		public string GetDescription() {
			return "Fix one incorrect Master Quartz effect description.";
		}

		public IEnumerable<FileModResult> TryApply(FileStorage storage) {
			var file_en = storage.TryGetDuplicate(new HyoutaUtils.Checksum.SHA1(0x40ae4f525021b550ul, 0xb77e045a00841a42ul, 0xbc460a77u));
			if (file_en == null) {
				return null;
			}

			var tbl_en = new Tbl(file_en, EndianUtils.Endianness.LittleEndian);

			// Katze says attack/craft instead of magic for its second effect
			{
				int idx = 323;
				var m = new MasterQuartzMemo(tbl_en.Entries[idx].Data);
				m.str = m.str.ReplaceSubstring(1, 14, new MasterQuartzMemo(tbl_en.Entries[325].Data).str, 1, 14);
				tbl_en.Entries[idx].Data = m.ToBinary();
			}

			Stream result_en = new MemoryStream();
			tbl_en.WriteToStream(result_en, EndianUtils.Endianness.LittleEndian);
			return new FileModResult[] { new FileModResult("data/text/dat_en/t_mstqrt.tbl", result_en) };
		}

		public IEnumerable<FileModResult> TryRevert(FileStorage storage) {
			var file_en = storage.TryGetDuplicate(new HyoutaUtils.Checksum.SHA1(0x40ae4f525021b550ul, 0xb77e045a00841a42ul, 0xbc460a77u));
			if (file_en == null) {
				return null;
			}
			return new FileModResult[] { new FileModResult("data/text/dat_en/t_mstqrt.tbl", file_en) };
		}
	}
}

File: native/sen4/file_fixes/t_mstqrt.cpp
#include <filesystem>
#include <string_view>
#include <vector>

#include "file_getter.h"

#include "p3a/pack.h"
#include "p3a/structs.h"
#include "sen4/tbl.h"
#include "sha1.h"
#include "util/text.h"

namespace SenLib::Sen4::FileFixes::t_mstqrt {
std::string_view GetDescription() {
    return "Fix one incorrect Master Quartz effect description.";
}

bool TryApply(const SenPatcher::GetCheckedFileCallback& getCheckedFile,
              std::vector<SenPatcher::P3APackFile>& result) {
    using namespace HyoutaUtils::TextUtils;

    try {
        auto file_en = getCheckedFile(
            "data/text/dat_en/t_mstqrt.tbl",
            40683,
            SenPatcher::SHA1FromHexString("40ae4f525021b550b77e045a00841a42bc460a77"));
        if (!file_en) {
            return false;
        }

        auto& bin = file_en->Data;

        Tbl tbl_en(bin.data(), bin.size(), HyoutaUtils::EndianUtils::Endianness::LittleEndian);

        // Katze says attack/craft instead of magic for its second effect
        {
            auto& e = tbl_en.Entries[323];
            const auto& e2 = tbl_en.Entries[325];
            MasterQuartzMemo m(e.Data.data(), e.Data.size());
            m.str = ReplaceSubstring(
                m.str, 1, 14, MasterQuartzMemo(e2.Data.data(), e2.Data.size()).str, 1, 14);
            e.Data = m.ToBinary();
        }

        std::vector<char> result_en_vec;
        MemoryStream result_en(result_en_vec);
        tbl_en.WriteToStream(result_en, HyoutaUtils::EndianUtils::Endianness::LittleEndian);

        result.emplace_back(
            std::move(result_en_vec), file_en->Filename, SenPatcher::P3ACompressionType::LZ4);

        return true;
    } catch (...) {
        return false;
    }
}
} // namespace SenLib::Sen4::FileFixes::t_mstqrt
